scipy/stats/tests/test_distributions.py

```diff
@@ -5297,6 +5296,19 @@ def test_sf_isf(self):
         x1 = stats.nakagami.isf(sf, nu)
         assert_allclose(x1, x0, rtol=1e-13)
 
+    @pytest.mark.parametrize('nu', [1.6, 2.5, 3.9])
+    @pytest.mark.parametrize('loc', [25.0, 10, 35])
+    @pytest.mark.parametrize('scale', [13, 5, 20])
+    def test_fit(self, nu, loc, scale):
```

Comment:
I think we could make this test more sensitive by checking that the likelihood equations are satisfied (those partial derivatives are nearly zero). Do you agree?

I think it would also be a good idea to check the special case in which `'nu' = 0.5` and for which the `fit` method fixes the parameter at that value, e.g. `stats.nakagami.fit(samples, f0=0.5)`. In this case, the likelihood equation for location will _not_ be satisfied, rather, you would check that `loc = np.min(samples)`. 

I think we're getting really close here. With these changes we've discussed, I think we can merge this.

Reply:
> I think we could make this test more sensitive by checking that the likelihood equations are satisfied (those partial derivatives are nearly zero). Do you agree?

I do agree. Generally, such checks could be added for ML estimators of distributions without constraints on their supports.
This is just another idea for gh-11782.

> I think it would also be a good idea to check the special case in which 'nu' = 0.5

I'll add an additional test with a comment explaining the motives you gave.

Reply:
OK, let me know when you'd like me to take a look!

Reply:
> In this case, the likelihood equation for location will not be satisfied, rather, you would check that `loc = np.min(samples)`.

Generally, we need to verify the constraint set on μ, which is not yet taken into account.
What if we introduce a new objective function combining the nnlf and a log barrier for the constraint on μ?

Reply:
> I'll add an additional test with a comment explaining the motives you gave.
Done in c3683c3.

Reply:
I think the optimization already does approximately what we want it to do here.`fit` already has a barrier of sorts... improving it is beyond the scope of this PR. (@WarrenWeckesser have been looking into that separately.)

Reply:
On some tests the derivatives aren't all null; I am inspecting it now before pushing modifications of the test.

Reply:
You'll have to decide whether the difference is significant or not. For example, here is the expression for the partial derivative with respect to location.
![image](https://user-images.githubusercontent.com/6570539/105321404-44023600-5b7c-11eb-80d2-6d8844df52bb.png)

Even if the derivative is not very small, is it small relative to each of those two terms? Do they mostly cancel each other out?

If they are _not_ canceling each other out, we need to look a little more closely. I would suggest trying to solve the system of equations so that they _are_ all zero, and then see if the log-likelihood function:
```python3
data = ...
def ll(v, u, s):
    return np.sum(np.log(nakagami.pdf(data, v, u, s)))
```
is larger with the parameters that do zero the partial derivatives than it is with the solution returned by `fit`. 

If `fit` gives us a larger log-likelihood than the solution to our system of equations, then something is wrong with our equations. If it's the other way around, then `fit` is not converging to the optimal solution.

I'm not so concerned with the latter. Numerical optimization isn't guaranteed to achieve the optimal solution. All we're trying to do in this PR is improve the guess; we still can't guarantee that the optimizer will find the optimal solution.

If the problem is our system of equations, we'll want to try to fix the error in the tutorial before we merge.

Reply:
In the case of the partial derivative of the likelihood w.r.t μ, since there's a constraint on its support (that is μ ≤ min_i x_i), it will hardly be equal to zero (i.e. the feasible optimal value for μ is not the optimum for the log-likelihood). I can observe that on most tests. For the two others partial derivatives, they are absolutely close to 0 up to 1e-3.

As for the system resolution, I have found that the log-likelihood associated to the solutions obtained with `scipy.optimize.root` (with Levenberg-Marquardt, for e.g) is less than the one of the solution obtains with `fit` for most tests (which is good!). For the others tests, the optimization routine/scheme of `scipy.optimize.root` yields negative values for ν (as the constraints have not been specified) which make the test fail. See [this gist](https://gist.github.com/jjerphan/4bc8e025919cbfcf1eac86635f98ccb1) for reproducibility.

As you said, improving the numerical optimisation routine is out of the scope of this PR. Though I am curious to have a look at own constraints of supports for parameters are or can be handle for `rv_continuous` subclasses' MLE/`fit` generally.

---

`tl;dr`: I've augmented the test with a check on the necessary condition for optimality on two of the partial derivatives in 72fa639b43f6a103b71087bce12d9fd955978d06.

Reply:
> In the case of the partial derivative of the likelihood w.r.t μ, since there's a constraint on its support (that is μ ≤ min_i x_i), it will hardly be equal to zero (i.e. the feasible optimal value for μ is not the optimum for the log-likelihood). I can observe that on most tests.

Well, unless something else is going, either
* the partial derivative should be zero or 
* the constraint should be active (that is, μ ≤ min_i x_i). 

We see the constraint active when nu is really close to 0.5. For all the other cases, the partial derivative looks nearly zero to me:

```python3
nus= [1.6, 2.5, 3.9]
locs = [25.0, 10, 35]
scales = [13, 5, 20]

for nu, loc, scale in product(nus, locs, scales):
    N = 100
    samples = stats.nakagami.rvs(size=N, nu=nu, loc=loc,
                                  scale=scale, random_state=1337)
    v, u, s = stats.nakagami.fit(samples)

    x = samples
    dldu = (-2*v + 1) * np.sum(1/(x - u)) + 2*v/s**2 * np.sum(x-u)
    print(dldu)
```

Reply:
> For all the other cases, the partial derivative looks nearly zero to me:

When I run this modified version of the script (with imports and printing):

```python
import numpy as np
from scipy import stats
from itertools import product
nus= [1.6, 2.5, 3.9]
locs = [25.0, 10, 35]
scales = [13, 5, 20]
for nu, loc, scale in product(nus, locs, scales):
     N = 100
     samples = stats.nakagami.rvs(size=N, nu=nu, loc=loc,
                                   scale=scale, random_state=1337)
     v, u, s = stats.nakagami.fit(samples)
 
     x = samples
     dldu = (-2*v + 1) * np.sum(1/(x - u)) + 2*v/s**2 * np.sum(x-u)
     print(f"(nu,loc,scale)=({nu}, {loc}, {scale}) \t\t dldu={dldu}")
     print(f"(v,u,s)=({v:.2f}, {u:.2f}, {s:.2f}) \t\t min(sample)={np.min(samples)}")
     print()
```
in most cases, I do not get a partial derivative which is null:
```
(nu,loc,scale)=(1.6, 25.0, 13) 		 dldu=267187821021946.56
(v,u,s)=(0.03, 26.86, 1.91) 		 min(sample)=26.862845727959257

(nu,loc,scale)=(1.6, 25.0, 5) 		 dldu=4.9108531364083774e-05
(v,u,s)=(1.74, 24.83, 5.26) 		 min(sample)=25.716479126138175

(nu,loc,scale)=(1.6, 25.0, 20) 		 dldu=136749576269998.34
(v,u,s)=(0.01, 27.87, 2.09) 		 min(sample)=27.8659165045527

(nu,loc,scale)=(1.6, 10, 13) 		 dldu=528323667636744.5
(v,u,s)=(0.03, 11.86, 2.00) 		 min(sample)=11.862845727959257

(nu,loc,scale)=(1.6, 10, 5) 		 dldu=200508511174458.75
(v,u,s)=(0.14, 10.72, 1.71) 		 min(sample)=10.716479126138175

(nu,loc,scale)=(1.6, 10, 20) 		 dldu=547515423430104.6
(v,u,s)=(0.01, 12.87, 2.14) 		 min(sample)=12.865916504552702

(nu,loc,scale)=(1.6, 35, 13) 		 dldu=134367465776150.02
(v,u,s)=(0.02, 36.86, 1.68) 		 min(sample)=36.862845727959254

(nu,loc,scale)=(1.6, 35, 5) 		 dldu=96135730775175.7
(v,u,s)=(0.16, 35.72, 1.84) 		 min(sample)=35.716479126138175

(nu,loc,scale)=(1.6, 35, 20) 		 dldu=137355339523805.89
(v,u,s)=(0.01, 37.87, 1.94) 		 min(sample)=37.8659165045527

(nu,loc,scale)=(2.5, 25.0, 13) 		 dldu=261796801631422.75
(v,u,s)=(0.03, 28.33, 1.69) 		 min(sample)=28.32581019355903

(nu,loc,scale)=(2.5, 25.0, 5) 		 dldu=194977126600637.5
(v,u,s)=(0.15, 26.28, 1.50) 		 min(sample)=26.27915776675347

(nu,loc,scale)=(2.5, 25.0, 20) 		 dldu=275506932660216.2
(v,u,s)=(0.01, 30.12, 1.67) 		 min(sample)=30.11663106701389

(nu,loc,scale)=(2.5, 10, 13) 		 dldu=504184655175108.25
(v,u,s)=(0.05, 13.33, 2.16) 		 min(sample)=13.325810193559029

(nu,loc,scale)=(2.5, 10, 5) 		 dldu=283985415051308.0
(v,u,s)=(0.25, 11.28, 1.83) 		 min(sample)=11.279157766753473

(nu,loc,scale)=(2.5, 10, 20) 		 dldu=8.439969640505751e-06
(v,u,s)=(2.77, 9.05, 21.20) 		 min(sample)=15.11663106701389

(nu,loc,scale)=(2.5, 35, 13) 		 dldu=134522106712885.08
(v,u,s)=(0.02, 38.33, 1.51) 		 min(sample)=38.32581019355903

(nu,loc,scale)=(2.5, 35, 5) 		 dldu=0.0002123006828327334
(v,u,s)=(2.77, 34.76, 5.30) 		 min(sample)=36.27915776675347

(nu,loc,scale)=(2.5, 35, 20) 		 dldu=136948858351787.02
(v,u,s)=(0.01, 40.12, 1.80) 		 min(sample)=40.11663106701389

(nu,loc,scale)=(3.9, 25.0, 13) 		 dldu=264856304242045.47
(v,u,s)=(0.03, 29.85, 1.51) 		 min(sample)=29.847553943506007

(nu,loc,scale)=(3.9, 25.0, 5) 		 dldu=134996579286915.28
(v,u,s)=(0.26, 26.86, 1.54) 		 min(sample)=26.864443824425386

(nu,loc,scale)=(3.9, 25.0, 20) 		 dldu=135290418754594.25
(v,u,s)=(0.02, 32.46, 1.83) 		 min(sample)=32.45777529770155

(nu,loc,scale)=(3.9, 10, 13) 		 dldu=-1.2802542002532391e-05
(v,u,s)=(4.21, 9.53, 13.61) 		 min(sample)=14.847553943506009

(nu,loc,scale)=(3.9, 10, 5) 		 dldu=339393765797140.2
(v,u,s)=(0.20, 11.86, 1.43) 		 min(sample)=11.864443824425388

(nu,loc,scale)=(3.9, 10, 20) 		 dldu=269159753937960.53
(v,u,s)=(0.02, 17.46, 2.11) 		 min(sample)=17.457775297701552

(nu,loc,scale)=(3.9, 35, 13) 		 dldu=127868092068877.98
(v,u,s)=(0.05, 39.85, 1.84) 		 min(sample)=39.84755394350601

(nu,loc,scale)=(3.9, 35, 5) 		 dldu=75980042179752.31
(v,u,s)=(0.23, 36.86, 1.61) 		 min(sample)=36.864443824425386

(nu,loc,scale)=(3.9, 35, 20) 		 dldu=134193438617571.53
(v,u,s)=(0.02, 42.46, 1.84) 		 min(sample)=42.45777529770155

```

> the constraint should be active (that is, μ ≤ min_i x_i).

I think that it is rather μ = min_i x_i if the constraint is active; see 4.1.1 (p. 127) of [Stephen Boyd and Lieven Vandenberghe. 2004. Convex Optimization. Cambridge University Press, USA.](https://web.stanford.edu/~boyd/cvxbook/bv_cvxbook.pdf) for reference.

The (quasi-)equality is also something that we tend to observe when the partial derivative is not null.

Reply:
Weird... I run that script and get all nearly zero! For example, the first few outputs:
```
(nu,loc,scale)=(1.6, 25.0, 13) 		 dldu=-5.5426409719672165e-05
(v,u,s)=(1.74, 24.55, 13.66) 		 min(sample)=26.862845727959257

(nu,loc,scale)=(1.6, 25.0, 5) 		 dldu=-0.00036703005109472997
(v,u,s)=(1.74, 24.83, 5.26) 		 min(sample)=25.716479126138175

(nu,loc,scale)=(1.6, 25.0, 20) 		 dldu=-1.5228397325373066e-05
(v,u,s)=(1.74, 24.30, 21.02) 		 min(sample)=27.8659165045527

(nu,loc,scale)=(1.6, 10, 13) 		 dldu=-1.1207026602733094e-05
(v,u,s)=(1.74, 9.55, 13.66) 		 min(sample)=11.862845727959257
```

Reply:
> I think that it is rather μ = min_i x_i if the constraint is active

Yes, I copy-pasted and forgot to change it to an equality sign, which is what active means.

Reply:
Hum, this is weird. Here are the versions that I use:
 - Python 3.9.1
 - scipy: '1.7.0.dev0+04a36f7' (this fork)
 - numpy: '1.19.4'

Reply:
My bad: I just was on another branch when executing, your results are right.

Reply:
Now in case you'd like to see a case where the optimizer _doesn't_ converge to the right answer, change `random_state` to 1...

Reply:
I guess 3b467ef5362e806669b9a61da6390108a35fbbff closes this thread. 😁